R/detect_mistakes.R

```diff
@@ -1,281 +1,302 @@
+  if (length(unmatched_user_names) > 0) {
+    surplus_name <- unmatched_user_names[1]
+    return(
+      surplus_argument(
+        this_call = user,
+        this = user[[surplus_name]],
+        this_name = surplus_name,
+        enclosing_call = enclosing_call,
+        enclosing_arg = enclosing_arg
+      )
+    )
+  }
+  
+  # 7. Check that every named argument in the solution matches every
+  #    correspondingly named argument in the user code. We know each 
+  #    has a match because of Step 5.
+  user_args <- as.list(user[-1])         # remove the call
+  solution_args <- as.list(solution[-1]) # remove the call
+  
+  for (name in actual_solution_names) {
+    if (!identical(user[[name]], solution[[name]])) {
+      arg_name <- ifelse(name %in% submitted_names, name, "")
+      return(
+        detect_mistakes(
+          user = user[[name]], 
+          solution = solution[[name]], 
+          env = env,
+          # If too verbose, use user[1]
+          enclosing_call = submitted,
+          # avoid naming first arguments in messages
+          enclosing_arg = arg_name
         )
+      )
+    }
+    
+    # Make these arguments invisible to further checks
+    user_args[[name]] <- NULL
+    solution_args[[name]] <- NULL
+  }
+  
+  
+  # 8. Extract the remaining arguments from the user code and the solution code.
+  #    Pair them in the order that they occur, checking that each pair matches.
+  #    Check pairs in sequence and address unmatched arguments when you get to
+  #    them.
+  user_len <- length(user_args)
+  solution_len <- length(solution_args)
+
+  n <- max(user_len, solution_len)
+  
+  for (i in seq_len(n)) {
+    
+    # if solution argument is unmatched due to no remaining user arguments
+    if (i > user_len) {
+      name <- rlang::names2(solution_args[i])
+      
+      # if the missing argument is unnamed, pass the value
+      if (is.null(name) || name == "") {
+        name <- solution_args[[i]]
       }
       
+      return(
+        missing_argument(
+          this_call = solution,
+          that_name = name,
+          enclosing_call = enclosing_call,
+          enclosing_arg = enclosing_arg
+        )
+      )
+      
+    # if user argument is unmatched due to no remaining solution arguments
+    } else if (i > solution_len) {
+      arg_name <- rlang::names2(user_args[i])
+      if (!(arg_name %in% submitted_names)) arg_name <- ""
+      return(
+        surplus_argument(
+          this_call = user,
+          this = user_args[[i]],
+          this_name = arg_name,
+          enclosing_call = enclosing_call,
+          enclosing_arg = enclosing_arg
+        )
+      )
+      
+    # The user argument has a matching solution argument, are they identical?
+    } else if (!identical(user_args[[i]], solution_args[[i]])) {
+      name <- rlang::names2(user_args[i])
+      if (!(name %in% submitted_names)) name <- ""
+      return(
```

Comment:
Only return if the result is `!is.null(res)`.  This applies to Section 7 above. ~line 217

Reply:
Have the first pair of unnamed args match. Have the second pair of unstandardized args mismatch.  Currently `detect_mistakes` will say _valid_. It should be _invalid_.

Reply:
Fixed in https://github.com/rstudio-education/gradethis/pull/92/commits/e4e8efc50e5c316314a05f881e09313c36dffee9